### ISSUE TITLE
Tooltip Position Wrong when the chart is in a dialog

### ISSUE THREAD

**Issue by rockydd** [novus/nvd3#1532]
The tooltip will have a wrong offset when the chart is in a bootstrap dialog. This issue is found in 1.8.2. While It works well in 1.8.1.

<img src="http://i.imgur.com/xVdrL1J.jpg?1">

**Comment by rockydd** [comment]
OK, found the root cause. Now the tooltip was translated based on the position returned by function position() use d3.event.clientX. However, when the parent container has 'transform' style, then the start point for the fixed position will be the left-top corner of the parent container instead of the browser window.

**Comment by laxu** [comment]
Same issue happens when using Foundation dialogs. Is there anything we can do to fix this until a new version is out aside from forking the project and changing the code there?

**Comment by smills2929** [comment]
I have the same issue as I am simply using translations on the chart containers. Has anyone figured out a work-around?

**Comment by liquidpele** [comment]
Can you submit a jsfiddle that demonstrates the problem, and link to the latest version of nvd3 in the master branch using rawgit.com ?

**Comment by smills2929** [comment]
Here you go. Works fine with a translate on body in 1.8.1. Is broken on 1.8.2 or latest commit.
1.8.1: https://jsfiddle.net/oxvygewj/
latest commit (c2ea2cb): https://jsfiddle.net/yxn42srr/

**Comment by nickcoutsos** [comment]
Is this or a related issue being looked at? I just spent an hour trying to clean up some CSS hacks my boss used to work around this issue and it seems like the alternative is just downgrading to 1.8.1, except apparently 1.8.2 introduces a fix for something else.

**Comment by liquidpele** [comment]
Updated jsfiddle to show latest master code, still looks to be an issue.  

https://jsfiddle.net/yxn42srr/1/

**Comment by nickcoutsos** [comment]
I did some digging in the history for the tooltip module, and the biggest red flag was #1190. There are a lot of changes for a newcomer to try to make sense of, but (if I've done this correctly), here's are jsfiddles using builds closest to that merge:

* [before](https://jsfiddle.net/ub1ranrr/2/) (3ba48c02b96d5fdc395beb0cd36102b271a2bf0c)
* [after](https://jsfiddle.net/ub1ranrr/1/) (fbdcb94c7b54fd22f52e7d354f67526af8a1eb05)

There were actually a bunch of commits between the builds but hopefully that can narrow things down for interested contributors. Looking at the diff from that PR the most significant difference to me seems to be based around the removal of the `tooltipTop`/`tooltipLeft` functions which worked to build up offsets of the tooltip's ancestors. 
https://github.com/RenatoUtsch/nvd3/blob/c8257cc4339725318a0898202ddf833369a28e55/src/tooltip.js#L169-L188

**Comment by liquidpele** [comment]
@RenatoUtsch looks like maybe this is a bug from you refactoring tooltip?

**Comment by RenatoUtsch** [comment]
Yeah, probably. I remember that the `tooltipTop` / `tooltipLeft` functions were pretty bugged and returned wrong results in many cases, so I used the d3 functions to get the position of the mouse relative to the window.

What I did not know is that the 'transform' attribute changes the start point for fixed position. Honestly, I think it would be easier (and better) to put the tooltip div directly in the document.body, like it was before. It would save us from problems like this. However, I remember, @liquidpele, that it was you who changed this when you rewrote the tooltip code. Do you remember why you changed the tooltip container from the body to the parent of the svg?

Another solution would be to always go up the hierarchy in the `position()` function and try to find the correct position from the `translate` transform of the parents. Actually, maybe this would always be needed at least for the document.body (if we put the tooltip directly in the body) if it has a translate transformation.

What do you think?

**Comment by RenatoUtsch** [comment]
Btw, this should be added to the unit tests.

**Comment by RenatoUtsch** [comment]
Everyone, please take a look at this PR and see if you can find any problems with it (it should probably fix all your problems): #1674

**Comment by RumenTech** [comment]
For those who can not upgrade nvd3 version, it seems to be the solution to reset modal container's `transform` style.

```
var containerModal = angular.element(svgElement).closest('.modal-dialog')[0];
if (containerModal && containerModal.style) {
  containerModal.style.transform = 'none';
}
```
